src/test/ui/generic-associated-types/issue-78113-lifetime-mismatch-dyn-trait-box.stderr

```diff
@@ -0,0 +1,34 @@
+error: incompatible lifetime on type
+  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:5
+   |
+LL |     type T<'a> = Box<dyn A + 'a>;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: because of this predicate
+  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:12:17
+   |
+LL |     type T<'a>: A;
+   |                 ^
+note: a compatible `impl` exists, but...
+  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:9:1
+   |
+LL | impl A for Box<dyn A> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^
+note: the lifetime `'a` as defined on the associated item at 16:12...
+  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:12
+   |
+LL |     type T<'a> = Box<dyn A + 'a>;
+   |            ^^
+   = note: ...does not necessarily outlive the static lifetime
+note: the used `impl` has a `'static` requirement
+  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:9:20
+   |
+LL | impl A for Box<dyn A> {}
+   |                    ^ this has an implicit `'static` lifetime requirement
+help: consider relaxing the implicit `'static` requirement
+   |
+LL | impl A for Box<dyn A + '_> {}
+   |                      ^^^^
```

Comment:
Given that we already have the whole thing available, it'll be interesting to try and craft the error to reduce its verbosity. What do you think of

```
error: incompatible lifetime on type
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:5
   |
LL |     type T<'a> = Box<dyn A + 'a>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: because this has an unmet lifetime requirement
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:12:17
   |
LL |     type T<'a>: A;
   |                 ^ introduces a `'static` lifetime requirement
note: a compatible `impl` exists, but...
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:9:1
   |
LL | impl A for Box<dyn A> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^
note: ...the lifetime `'a` as defined on the associated item at 16:12...
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:12
   |
LL |     type T<'a> = Box<dyn A + 'a>;
   |            ^^
note: ...does not necessarily outlive the static lifetime introduced by the `impl`
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:9:20
   |
LL | impl A for Box<dyn A> {}
   |                    ^ this has an implicit `'static` lifetime requirement
help: consider relaxing the implicit `'static` requirement
   |
LL | impl A for Box<dyn A + '_> {}
   |                      ^^^^
```

or

```
error: incompatible lifetime on type
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:5
   |
LL |     type T<'a> = Box<dyn A + 'a>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incompatible lifetimes
...
LL |     type T<'a>: A;
   |                 - introduces a `'static` lifetime requirement
note: the lifetime `'a` as defined on the associated item at 16:12...
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:16:12
   |
LL |     type T<'a> = Box<dyn A + 'a>;
   |            ^^
note: ...does not necessarily outlive the static lifetime introduced by the compatible `impl`
  --> $DIR/issue-78113-lifetime-mismatch-dyn-trait-box.rs:9:20
   |
LL | impl A for Box<dyn A> {}
   |                    ^ this has an implicit `'static` lifetime requirement
help: consider relaxing the implicit `'static` requirement
   |
LL | impl A for Box<dyn A + '_> {}
   |                      ^^^^
```

Doing this requires fiddling with the logic in `mismatched_static_lifetime.rs` to accomplish something like these.

Reply:
I like the second one. The only problem with point at the `A` on the associated type bound and saying that it introduces a 'static lifetime requirement is that it's not a requirement for the whole type, only for that trait object. While this won't be a problem *here* (because we won't imply a 'static lifetime when the type itself is shorter), I could imagine in the future extending this to also include a suggestion for explicit 'static annotations:

Example of a test where we could provide an additional suggestion in the future:
```rust
pub trait A {}
impl<'a> A for &'a (dyn A + 'static) {}  // here the type itself has a lifetime of `'a`, but the trait object has a lifetime of `'static`
pub trait B {
    type T<'a>: A;
}
impl B for () {
    type T<'a> = &'a (dyn A + 'a);
}
```

Reply:
How's it look now?

Reply:
Looks great!